modules/videoio/src/videoio_registry.cpp

```diff
@@ -0,0 +1,640 @@
+        return g_instance;
+    }
+
+    inline std::vector<VideoBackendInfo> getAvailableBackends_CaptureByIndex() const
+    {
+        std::vector<VideoBackendInfo> result;
+        for (size_t i = 0; i < enabledBackends.size(); i++)
+        {
+            const VideoBackendInfo& info = enabledBackends[i];
+            if (info.mode & MODE_CAPTURE_BY_INDEX)
+                result.push_back(info);
+        }
+        return result;
+    }
+    inline std::vector<VideoBackendInfo> getAvailableBackends_CaptureByFilename() const
+    {
+        std::vector<VideoBackendInfo> result;
+        for (size_t i = 0; i < enabledBackends.size(); i++)
+        {
+            const VideoBackendInfo& info = enabledBackends[i];
+            if (info.mode & MODE_CAPTURE_BY_FILENAME)
+                result.push_back(info);
+        }
+        return result;
+    }
+    inline std::vector<VideoBackendInfo> getAvailableBackends_Writer() const
+    {
+        std::vector<VideoBackendInfo> result;
+        for (size_t i = 0; i < enabledBackends.size(); i++)
+        {
+            const VideoBackendInfo& info = enabledBackends[i];
+            if (info.mode & MODE_WRITER)
+                result.push_back(info);
+        }
+        return result;
+    }
+};
+
+} // namespace
+
+namespace videoio_registry {
+
+std::vector<VideoBackendInfo> getAvailableBackends_CaptureByIndex()
+{
+    const std::vector<VideoBackendInfo> result = VideoBackendRegistry::getInstance().getAvailableBackends_CaptureByFilename();
+    return result;
+}
+std::vector<VideoBackendInfo> getAvailableBackends_CaptureByFilename()
+{
+    const std::vector<VideoBackendInfo> result = VideoBackendRegistry::getInstance().getAvailableBackends_CaptureByFilename();
+    return result;
+}
+std::vector<VideoBackendInfo> getAvailableBackends_Writer()
+{
+    const std::vector<VideoBackendInfo> result = VideoBackendRegistry::getInstance().getAvailableBackends_Writer();
+    return result;
+}
+
+} // namespace registry
+
+#define TRY_OPEN(backend_func) \
+{ \
+    try { \
+        if (param_VIDEOIO_DEBUG || param_VIDEOCAPTURE_DEBUG) \
+            CV_LOG_WARNING(NULL, cv::format("VIDEOIO(%s): trying ...\n", #backend_func)); \
+        icap = backend_func; \
+        if (param_VIDEOIO_DEBUG ||param_VIDEOCAPTURE_DEBUG) \
+            CV_LOG_WARNING(NULL, cv::format("VIDEOIO(%s): result=%p isOpened=%d ...\n", \
+                #backend_func, icap.empty() ? NULL : icap.get(), icap.empty() ? -1: icap->isOpened())); \
+    } catch(const cv::Exception& e) { \
+        CV_LOG_ERROR(NULL, cv::format("VIDEOIO(%s): raised OpenCV exception:\n\n%s\n", #backend_func, e.what())); \
+    } catch (const std::exception& e) { \
+        CV_LOG_ERROR(NULL, cv::format("VIDEOIO(%s): raised C++ exception:\n\n%s\n", #backend_func, e.what())); \
+    } catch(...) { \
+        CV_LOG_ERROR(NULL, cv::format("VIDEOIO(%s): raised unknown C++ exception!\n\n", #backend_func)); \
+    } \
+    break; \
```

Comment:
IMHO it would be better to remove unconditional brake here and add it explicitly after every macro call. Explicit brake is already added after almost every `TRY_OPEN`, `TRY_OPEN_LEGACY`. So for a few places without explicit brake it could be unclear that processing will be stopped on first TRY.

Reply:
"break" should be unconditional by design.
I added "break"s to make happy IDE (Eclipse CDT). Missing breaks are errors - I will re-check that again.

Reply:
VFW shares **the same** api id with V4L/V4L2 (and can't be used simultaneously). Also there is no way to use multiple FIREWIRE backends - only one can be selected.